src/main/java/ru/l1ttleO/boyfriend/Utils.java

```diff
@@ -16,81 +28,195 @@ public static String getBeep() {
+                millisOfTheDay /= max;
+            } else {
+                if (dateFrom.get(unit) == dateTo.get(unit))
+                    continue;
+                switch (unit) {
+                    case Calendar.YEAR:
+                        amount = dateTo.get(unit) - dateFrom.get(unit);
                         break;
-                    case 1:
-                        out.insert(0, " и ");
+                    case Calendar.MONTH:
+                        amount = dateTo.get(Calendar.MONTH) - dateFrom.get(Calendar.MONTH);
+                        if (amount < 0) {
+                            amount += 12;
+                            durations.put(Calendar.YEAR, durations.get(Calendar.YEAR) - 1);
+                        }
                         break;
-                    default:
-                        out.insert(0, ", ");
+                    default: // DAY_OF_YEAR
+                        amount = dateTo.get(Calendar.DAY_OF_MONTH) - dateFrom.get(Calendar.DAY_OF_MONTH);
+                        if (amount < 0) {
+                            amount = durations.getOrDefault(Calendar.MONTH, 12);
+                            final int years = durations.getOrDefault(Calendar.YEAR, 0);
+                            if (amount == 12)
+                                durations.put(Calendar.YEAR, years - 1);
+                            durations.put(Calendar.MONTH, amount - 1);
+                            amount = dateTo.get(Calendar.DAY_OF_MONTH);
+                            dateTo.add(Calendar.MONTH, -1);
+                            max = dateTo.getActualMaximum(Calendar.DAY_OF_MONTH);
+                            amount += max - Math.min(max, dateFrom.get(Calendar.DAY_OF_MONTH));
+                        }
+                        durations.put(Calendar.WEEK_OF_YEAR, amount / 7);
+                        amount %= 7;
                         break;
                 }
-                concatLength++;
-                out.insert(0, amount + " " + plural(amount, DURATION_TEXTS[i][vin ? 2 : 1], DURATION_TEXTS[i][3], DURATION_TEXTS[i][4]));
-                seconds -= amount;
             }
-            seconds /= DURATION_MULTIPLIERS[i];
+            durations.put(unit, amount);
         }
+
+        // filter and sort
+        for (final int unit : DURATION_TYPES.keySet()) {
+            final int amount = durations.getOrDefault(unit, 0);
+            durations.remove(unit);
+            if (amount != 0)
+                durations.put(unit, amount);
+        }
+        
+        // build a string
+        final StringBuilder out = new StringBuilder();
+        int i = 0;
+        String[] names;
+        for (final var entry : durations.entrySet()) {
+            names = DURATION_TYPES.get(entry.getKey()).getRight();
+            out.append((negative && entry.getValue() > 0 ? "-" : "") + entry.getValue() + " " + plural(entry.getValue(), names[accusative ? 1 : 0], names[2], names[3]));
```

Comment:
```suggestion
            out.append(negative && entry.getValue() > 0 ? "-" : "").append(entry.getValue()).append(" ").append(plural(entry.getValue(), names[accusative ? 1 : 0], names[2], names[3]));
```
`String concatenation as argument to 'StringBuilder.append()' call`